Add Scatter op to ONNX

pk-g opened the pull request:
Adding Scatter op and corresponding tests to ONNX

spandantiwari commented on onnx/backend/test/case/node/scatter.py:
```diff
@@ -0,0 +1,51 @@
+class Scatter(Base):
+
+    @staticmethod
+    def export_scatter_0():  # type: () -> None
```
I would suggest using better names for these test points. For example, we can eliminate the explicit use of the `axis=0` attribute here, and call this `export_without_axis`. In the next test point, we can have the axis attribute (`axis=1`) and that can be called `export_with_axis`.

pk-g replied:
fixed

pk-g commented on onnx/defs/operator_sets.h:
```diff
@@ -93,6 +93,7 @@ class ONNX_OPERATOR_SET_SCHEMA_CLASS_NAME(Onnx, 1, Relu);
 class ONNX_OPERATOR_SET_SCHEMA_CLASS_NAME(Onnx, 1, Reshape);
 class ONNX_OPERATOR_SET_SCHEMA_CLASS_NAME(Onnx, 1, Scale);
 class ONNX_OPERATOR_SET_SCHEMA_CLASS_NAME(Onnx, 1, ScaledTanh);
+class ONNX_OPERATOR_SET_SCHEMA_CLASS_NAME(Onnx, 1, Scatter);
```
fixed

pk-g commented on onnx/defs/operator_sets.h:
```diff
@@ -225,6 +226,7 @@ class OpSet_Onnx_ver1 {
     fn(GetOpSchema<ONNX_OPERATOR_SET_SCHEMA_CLASS_NAME(Onnx, 1, Reshape)>());
     fn(GetOpSchema<ONNX_OPERATOR_SET_SCHEMA_CLASS_NAME(Onnx, 1, Scale)>());
     fn(GetOpSchema<ONNX_OPERATOR_SET_SCHEMA_CLASS_NAME(Onnx, 1, ScaledTanh)>());
+    fn(GetOpSchema<ONNX_OPERATOR_SET_SCHEMA_CLASS_NAME(Onnx, 1, Scatter)>());
```
fixed

pk-g commented on onnx/defs/tensor/defs.cc:
```diff
@@ -610,6 +610,94 @@ ONNX_OPERATOR_SET_SCHEMA(
           }
         }));
 
+static const char* Scatter_ver1_doc = R"DOC(
```
fixed

pk-g commented on onnx/defs/tensor/defs.cc:
```diff
@@ -610,6 +610,94 @@ ONNX_OPERATOR_SET_SCHEMA(
+
+ONNX_OPERATOR_SET_SCHEMA(
+    Scatter,
+    1,
```
fixed

pk-g commented on onnx/defs/tensor/defs.cc:
```diff
@@ -610,6 +610,94 @@ ONNX_OPERATOR_SET_SCHEMA(
+
+          int out_rank = r;
+
+          if (out_rank == 0) {
```
fixed.

pk-g commented on onnx/defs/tensor/defs.cc:
```diff
@@ -610,6 +610,94 @@ ONNX_OPERATOR_SET_SCHEMA(
+        .TypeConstraint(
+            "T",
+            OpSchema::all_tensor_types(),
+            "Constrain input and output types to any tensor type.")
```
fixed

houseroad commented on onnx/backend/test/case/node/scatter.py:
```diff
@@ -0,0 +1,50 @@
+            outputs=['y'],
+        )
+        data = np.zeros((3, 3))
+        indices = np.array([[1, 0, 2], [0, 2, 1]])
```
please specify the dtype as int64, otherwise, we may generate different type of data on 32bit platform.

pk-g replied:
sure, resolved. good point!

houseroad commented on onnx/backend/test/case/node/scatter.py:
```diff
@@ -0,0 +1,50 @@
+            outputs=['y'],
+            axis=1,
+        )
+        data = np.array([[1.0, 2.0, 3.0, 4.0, 5.0]])
```
ditto

pk-g replied:
resolved.

houseroad commented on onnx/defs/tensor/defs.cc:
```diff
@@ -610,6 +610,91 @@ ONNX_OPERATOR_SET_SCHEMA(
+            "Input and output types can be of any tensor type.")
+        .TypeConstraint(
+            "Tind",
+            {"tensor(int32)", "tensor(int64)"},
```
just int64? I think other ops only support int64 as index, not much sense to support int32 here.

pk-g replied:
since we are already supporting both int32 and int64 in gather(based on https://github.com/onnx/onnx/blob/master/docs/Operators.md#Gather), and given the fact that scatter can conceptually be viewed as opposite of gather, it may make sense for scatter to support what is already being supported by gather ?

houseroad commented on onnx/defs/tensor/defs.cc:
```diff
@@ -610,6 +610,91 @@ ONNX_OPERATOR_SET_SCHEMA(
+        .Input(
+            2,
+            "updates",
+            "Tensor of rank r >=1 (same rank and type as input data)",
```
Same shape as indices?

pk-g replied:
Sure, resolved.

pk-g commented:
@houseroad, thanks for the initial review. Can you please review the updated version so we could move forward with adding this op? thanks.

fdwr commented on docs/Changelog.md:
```diff
@@ -8873,3 +8873,76 @@ This version of the operator has been available since version 9 of the default O
+  into `data` along `axis` dimension of `data` (by default outer-most one as axis=0) at corresponding `indices`. 
+  For each entry in `updates`, the target index in `data` is specified by corresponding entry in `indices`
+  for dimension = axis, and index in source for dimension != axis. For instance, in a 2-D tensor case, 
+  data[indices[i][j]][j] = updates[i][j] if axis = 0, or data[i][indices[i][j]] = updates[i][j] if axis = 1.
```
Something that was unclear to me is that `i` and `j` are not loop counters in terms of `data` or `output` but rather `indices` and `updates`. This greatly bewildered me when looking at example 2 because it appeared that "data[i][indices[i][j]] = updates[i][j] if axis = 1" would yield out of bounds array accesses. Propose clarifying that...

```suggestion
  data[indices[i][j]][j] = updates[i][j] if axis = 0, or data[i][indices[i][j]] = updates[i][j] if axis = 1, where i and j are loop counters from 0 up to the respective size in `updates` - 1.
```

fdwr commented on onnx/defs/tensor/defs.cc:
```diff
@@ -610,6 +610,91 @@ ONNX_OPERATOR_SET_SCHEMA(
+into `data` along `axis` dimension of `data` (by default outer-most one as axis=0) at corresponding `indices`. 
+For each entry in `updates`, the target index in `data` is specified by corresponding entry in `indices`
+for dimension = axis, and index in source for dimension != axis. For instance, in a 2-D tensor case, 
+data[indices[i][j]][j] = updates[i][j] if axis = 0, or data[i][indices[i][j]] = updates[i][j] if axis = 1.
```
Something that was unclear to me is that `i` and `j` are not loop counters in terms of `data` or `output` but rather `indices` and `updates`. This greatly bewildered me when looking at example 2 because it appeared that "data[i][indices[i][j]] = updates[i][j] if axis = 1" would yield out of bounds array accesses. Propose clarifying that...

```suggestion
  data[indices[i][j]][j] = updates[i][j] if axis = 0, or data[i][indices[i][j]] = updates[i][j] if axis = 1, where i and j are loop counters from 0 up to the respective size in `updates` - 1.
```

pk-g replied:
sure, thanks for the proposal.

pk-g commented:
@houseroad, thanks again for the initial review and feedbacks. I have incorporated the feedback and updated the PR. can you please review the updated PR so we can move forward with adding this op ? thanks!

pk-g commented:
@houseroad thanks for the review, the CI is now passing, can you please take another look so we can move forward with adding this op ?

pk-g commented on onnx/defs/tensor/defs.cc:
```diff
@@ -606,6 +606,91 @@ ONNX_OPERATOR_SET_SCHEMA(
+            return;
+          }
+
+          int r = ctx.getInputType(0)->tensor_type().shape().dim_size();
```
That's right, fixed and added the shape inference tests.

pk-g commented:
@houseroad sure, updated the PR and applied feedback. can you please take a look so we can move forward ? thanks.

pk-g commented:
@houseroad just a friendly reminder here, can you please take a look at updated PR so we can move forward with adding this op? thanks.

houseroad commented on onnx/defs/tensor/defs.cc:
```diff
@@ -606,6 +606,81 @@ ONNX_OPERATOR_SET_SCHEMA(
+            "Constrain indices to integer types")
+        .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
+          propagateElemTypeFromInputToOutput(ctx, 0, 0);
+          if (hasNInputShapes(ctx, 3)) {
```
`hasNInputShapes(ctx, 1)`? Since you just need the shape information of the first input.

pk-g replied:
the other two inputs, indices and updates are not optional for this op. So we should always have three inputs when calling this op.

houseroad replied:
Yeah, they are required inputs. However, their shape information is not required for our shape inference. `hasNInputShapes` checks whether we have necessary information for us to do the shape inference.

pk-g replied:
makes sense, I'll update this to 1 then.

pk-g replied:
just updated the PR with this fix and resolved conflicts.
